---
title: "TheNessLab : Making an ideogram / karyogram"
---

::: {#page}
::: {#main .aui-page-panel}
::: {#main-header}
::: {#breadcrumb-section}
1.  [TheNessLab](index.html)
2.  [The Ness Lab Wiki](The-Ness-Lab-Wiki_11436042.html)
3.  [Coding tips and tutorials](Coding-tips-and-tutorials_11436186.html)
4.  [Making figures: ggplot examples](28016741.html)
:::

# [ TheNessLab : Making an ideogram / karyogram ]{#title-text} {#title-heading .pagetitle}
:::

::: {#content .view}
::: page-metadata
Created by [ Josianne Lachapelle]{.author} on Sep 05, 2018
:::

::: {#main-content .wiki-content .group}
The following code allows you to plot events according to their position
on each chromosome. For example, you might want to plot the position
of *de novo *mutations, or the position of TEs across the genome.

\

1.  You will need the following R packages:

```{=html}
<!-- -->
```
    library(GenomicRanges)
    library(ggbio)
    library(ggplot2)

\

\

2\. Create a data frame containing a column for each chromosome name,
and a column for the size of each chromosome.

    chr.data<-data.frame(chromosome = c(paste0('chromosome_',1:17), paste0('scaffold_',18:54), c('cpDNA','mtDNA', 'mtMinus')),
                         size = c(8033585, 9223677, 9219486, 4091191, 3500558, 9023763, 6421821, 5033832, 7956127,
                                  6576019, 3826814, 9730733, 5206065, 4157777, 1922860, 7783580, 7188315, 271631,
                                  219038, 200793, 189560, 163774, 127913, 127161, 102191, 80213, 55320, 55278,
                                  52813, 52376, 48183, 42264, 39192, 33576, 32450, 25399, 24537, 24437, 22408,
                                  22082, 21325, 21000, 20974, 17736, 16939, 16627, 14746, 14165, 13462, 12727,
                                  11225, 6241, 2479, 2277, 203828, 15758, 345555))

\

or in this case I only want to plot the chromosomes, not the scaffolds
or the plastids and so I use the following code:

    chr.data<-data.frame(chromosome = c(paste0('chromosome_',1:17)), 
                         size = c(8033585, 9223677, 9219486, 4091191, 3500558, 9023763, 6421821, 5033832, 7956127,
                                  6576019, 3826814, 9730733, 5206065, 4157777, 1922860, 7783580, 7188315))

\

\

3\. Import your data. Make sure your data is in the long format (as
shown below). At minimum, your data needs two columns: \'chromosome\',
and \'position\'. Make sure the levels of your \'**chromosome**\' factor
are the same as in the **chr.data**.

    all.mutations<-read.csv("my.mutation.data.csv", header=TRUE)

    all.mutations$chromosome<-factor(all.mutations$chromosome, levels=chr.data$chromosome)

    View(all.mutations)

::: table-wrap
<table class="confluenceTable" style="letter-spacing: 0.0px;">
<thead>
<tr class="header">
<th class="confluenceTh" style="text-align: right;"><br />
</th>
<th class="confluenceTh">Gene.primaryIdentifier</th>
<th class="confluenceTh">chromsome</th>
<th class="confluenceTh">position</th>
<th class="confluenceTh">mutation</th>
<th class="confluenceTh">type</th>
<th class="confluenceTh">mutant_sample</th>
<th class="confluenceTh">genic</th>
<th class="confluenceTh">exonic</th>
<th class="confluenceTh">intronic</th>
<th class="confluenceTh">intergenic</th>
<th class="confluenceTh">utr5</th>
<th class="confluenceTh">utr3</th>
<th class="confluenceTh">fold0</th>
<th class="confluenceTh">fold4</th>
<th class="confluenceTh">fold2</th>
<th class="confluenceTh">fold3</th>
<th class="confluenceTh">CDS</th>
<th class="confluenceTh">mRNA</th>
<th class="confluenceTh">rRNA</th>
<th class="confluenceTh">tRNA</th>
<th class="confluenceTh">FPKM</th>
<th class="confluenceTh">Nonsyn_V_Syn</th>
<th class="confluenceTh">X</th>
<th class="confluenceTh">nessID</th>
<th class="confluenceTh">FPKM_per_transcript</th>
<th class="confluenceTh">cluster</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td class="confluenceTd" style="text-align: right;"><p>1</p></td>
<td class="confluenceTd"><p>g88</p></td>
<td class="confluenceTd"><p>chromosome_1</p></td>
<td class="confluenceTd"><p>678032</p></td>
<td class="confluenceTd"><p>T&gt;A</p></td>
<td class="confluenceTd"><p>SNP</p></td>
<td class="confluenceTd"><p>./S26/</p></td>
<td class="confluenceTd"><p>1</p></td>
<td class="confluenceTd"><p>1</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>1</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>1</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0.457298757</p></td>
<td class="confluenceTd"><p>n/a</p></td>
<td class="confluenceTd"><p>265</p></td>
<td class="confluenceTd"><p>26904017</p></td>
<td class="confluenceTd"><p>4.572988e-01</p></td>
<td class="confluenceTd"><p>non-cluster</p></td>
</tr>
<tr class="even">
<td class="confluenceTd" style="text-align: right;"><p>2</p></td>
<td class="confluenceTd"><p>Cre01.g009750</p></td>
<td class="confluenceTd"><p>chromosome_1</p></td>
<td class="confluenceTd"><p>1815972</p></td>
<td class="confluenceTd"><p>A&gt;G</p></td>
<td class="confluenceTd"><p>SNP</p></td>
<td class="confluenceTd"><p>./S30/</p></td>
<td class="confluenceTd"><p>1</p></td>
<td class="confluenceTd"><p>1</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>1</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>1</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0.318886874</p></td>
<td class="confluenceTd"><p>n/a</p></td>
<td class="confluenceTd"><p>247</p></td>
<td class="confluenceTd"><p>26903734</p></td>
<td class="confluenceTd"><p>3.188869e-01</p></td>
<td class="confluenceTd"><p>non-cluster</p></td>
</tr>
<tr class="odd">
<td class="confluenceTd" style="text-align: right;"><p>3</p></td>
<td class="confluenceTd"><p>Cre01.g012700</p></td>
<td class="confluenceTd"><p>chromosome_1</p></td>
<td class="confluenceTd"><p>2318299</p></td>
<td class="confluenceTd"><p>G&gt;A</p></td>
<td class="confluenceTd"><p>SNP</p></td>
<td class="confluenceTd"><p>./S31/</p></td>
<td class="confluenceTd"><p>1</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>1</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>1</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>.</p></td>
<td class="confluenceTd"><p>n/a</p></td>
<td class="confluenceTd"><p>261</p></td>
<td class="confluenceTd"><p>26903906</p></td>
<td class="confluenceTd"><p>2.224498e+01</p></td>
<td class="confluenceTd"><p>non-cluster</p></td>
</tr>
<tr class="even">
<td class="confluenceTd" style="text-align: right;"><p>4</p></td>
<td class="confluenceTd"><p>Cre01.g012700</p></td>
<td class="confluenceTd"><p>chromosome_1</p></td>
<td class="confluenceTd"><p>2326644</p></td>
<td class="confluenceTd"><p>G&gt;A</p></td>
<td class="confluenceTd"><p>SNP</p></td>
<td class="confluenceTd"><p>./S21/</p></td>
<td class="confluenceTd"><p>1</p></td>
<td class="confluenceTd"><p>1</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>1</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>1</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>22.24497799</p></td>
<td class="confluenceTd"><p>n/a</p></td>
<td class="confluenceTd"><p>261</p></td>
<td class="confluenceTd"><p>26903906</p></td>
<td class="confluenceTd"><p>2.224498e+01</p></td>
<td class="confluenceTd"><p>non-cluster</p></td>
</tr>
<tr class="odd">
<td class="confluenceTd" style="text-align: right;"><p>5</p></td>
<td class="confluenceTd"><p>Cre01.g012900</p></td>
<td class="confluenceTd"><p>chromosome_1</p></td>
<td class="confluenceTd"><p>2355468</p></td>
<td class="confluenceTd"><p>T&gt;A</p></td>
<td class="confluenceTd"><p>SNP</p></td>
<td class="confluenceTd"><p>./S26/</p></td>
<td class="confluenceTd"><p>1</p></td>
<td class="confluenceTd"><p>1</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>1</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>1</p></td>
<td class="confluenceTd"><p>1</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0.876146117</p></td>
<td class="confluenceTd"><p>nonsynonymous</p></td>
<td class="confluenceTd"><p>257</p></td>
<td class="confluenceTd"><p>26903837</p></td>
<td class="confluenceTd"><p>8.761461e-01</p></td>
<td class="confluenceTd"><p>non-cluster</p></td>
</tr>
<tr class="even">
<td class="confluenceTd" style="text-align: right;"><p>6</p></td>
<td class="confluenceTd"><p>Cre01.g015200</p></td>
<td class="confluenceTd"><p>chromosome_1</p></td>
<td class="confluenceTd"><p>2615733</p></td>
<td class="confluenceTd"><p>A&gt;AC</p></td>
<td class="confluenceTd"><p>indel</p></td>
<td class="confluenceTd"><p>./S21/</p></td>
<td class="confluenceTd"><p>1</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>1</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>1</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>0</p></td>
<td class="confluenceTd"><p>.</p></td>
<td class="confluenceTd"><p>n/a</p></td>
<td class="confluenceTd"><p>266</p></td>
<td class="confluenceTd"><p>26904033</p></td>
<td class="confluenceTd"><p>2.583813e+00</p></td>
<td class="confluenceTd"><p>non-cluster</p></td>
</tr>
</tbody>
</table>
:::

\

\

4\. Make the plot. Here I\'m going to colour the mutations according to
whether the mutations are in genic regions or intergenic regions. You
can choose whichever factor interests you to distinguish different types
of events. You can also leave out the colours all together by removing
the last argument of the GRanges object.  

    all.muts.ranges<-GRanges(seqnames=all.mutations$chromosome,
                             IRanges(start=all.mutations$position,
                             width = rep(1,length(all.mutations$position))),
                             Genic=factor(all.mutations$genic))

    seqlengths(all.muts.ranges)<-chr.data$size

    pdf('my.mutations.karyogram.pdf')
    autoplot(all.muts.ranges, layout="karyogram", aes(colour = Genic))
    dev.off()

\

[![](attachments/28017247/28017245.png){.confluence-embedded-image
draggable="false" height="400"
image-src="attachments/28017247/28017245.png"
unresolved-comment-count="0" linked-resource-id="28017245"
linked-resource-version="1" linked-resource-type="attachment"
linked-resource-default-alias="image2018-9-5_16-2-4.png"
base-url="https://wiki.utm.utoronto.ca"
linked-resource-content-type="image/png"
linked-resource-container-id="28017247"
linked-resource-container-version="1"}]{.confluence-embedded-file-wrapper
.confluence-embedded-manual-size}

\

## [Taking it one step further]{.underline} {#Makinganideogram/karyogram-Takingitonestepfurther}

**Plotting events longer than one base**

If you want to plot blocks (such as haplotype blocks, or TEs, or indels)
that are longer than one base, you can change the **width** argument of
your GRanges object. In this particular example, I used the number
100,000 in my width argument because each block is 100,000 bases long.
The haplotype blocks are composed of one or more blocks of 100,000
bases, and their start and end positions is specified by the columns
\'Start.location\' and \'End.location\'.

    haplotype_ranges<-GRanges(seqnames=blocks.data$Chromosome,
                              IRanges(start=blocks.data$Start.location,
                              width=rep(100000, length(blocks.data$End.location))),
                              haplotype=blocks.data$Haplotype)

\

**Layering plots**

You can layer different karyograms. For example, you might want to layer
the haplotype blocks onto the single nucleotide mutations. To do so, you
create separate GRanges objects for each layer. And then you add them up
at the time of producing the plot.

\

    haplotype_ranges<-GRanges(seqnames=blocks.data$Chromosome,
                              IRanges(start=blocks.data$Start.location,
                                      width=rep(100000, length(blocks.data$End.location))),
                              haplotype=blocks.data$Haplotype)

    mut_ranges<-GRanges(seqnames=muts$chromosome,
                        IRanges(start=muts$position,
                                width = rep(1,length(muts$position))))

    pdf('my.layered.karyogram.pdf')
    autoplot(seqnames(haplotype_ranges), layout="karyogram") +
     layout_karyogram(data=haplotype_ranges, geom="rect", ylim=c(0, 10000), aes(fill=haplotype)) +
     layout_karyogram(data = mut_ranges, geom="rect", ylim=c(0,10000))
    dev.off()

[[![](rest/documentConversion/latest/conversion/thumbnail/28017246/1){height="400"
draggable="false"}](/download/attachments/28017247/S27.snps.on.haplotypes.pdf?version=1&modificationDate=1536179835000&api=v2){.confluence-embedded-file
nice-type="PDF Document"
file-src="/download/attachments/28017247/S27.snps.on.haplotypes.pdf?version=1&modificationDate=1536179835000&api=v2"
linked-resource-id="28017246" linked-resource-type="attachment"
linked-resource-container-id="28017247"
linked-resource-default-alias="S27.snps.on.haplotypes.pdf"
mime-type="application/pdf" has-thumbnail="true"
linked-resource-version="1" can-edit="true"
aria-label="S27.snps.on.haplotypes.pdf"
draggable="false"}[]{.companion-edit-button-placeholder
.edit-button-overlay linked-resource-container-id="28017247"
linked-resource-id="28017246" template-name="companionEditIcon"
source-location="embedded-attachment"}]{.confluence-embedded-file-wrapper}

\

\
:::

::: {.pageSection .group}
::: pageSectionHeader
## Attachments: {#attachments .pageSectionTitle}
:::

::: {.greybox align="left"}
![](images/icons/bullet_blue.gif){height="8" width="8"}
[image2018-9-5_16-2-4.png](attachments/28017247/28017245.png)
(image/png)\
![](images/icons/bullet_blue.gif){height="8" width="8"}
[S27.snps.on.haplotypes.pdf](attachments/28017247/28017246.pdf)
(application/pdf)\
:::
:::
:::
:::

::: {#footer role="contentinfo"}
::: {.section .footer-body}
Document generated by Confluence on May 22, 2024 11:44

::: {#footer-logo}
[Atlassian](https://www.atlassian.com/)
:::
:::
:::
:::
